### PR TITLE
[Core] put small objects in memory store

### PR DESCRIPTION
**Pull request by zhuohan123**
## Why are these changes needed?

Put objects that are smaller than `max_direct_call_object_size` into memory store instead of object store (plasma).

In the original code, put is sometimes used for putting objects in the object store store. We replace the put call in this case with codes that directly put objects into the object store.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -366,9 +366,26 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
       boost::bind(&CoreWorker::TriggerGlobalGC, this),
       boost::bind(&CoreWorker::CurrentCallSite, this)));
   memory_store_.reset(new CoreWorkerMemoryStore(
-      [this](const RayObject &obj, const ObjectID &obj_id) {
-        RAY_LOG(DEBUG) << "Promoting object to plasma " << obj_id;
-        RAY_CHECK_OK(Put(obj, /*contained_object_ids=*/{}, obj_id, /*pin_object=*/true));
+      [this](const RayObject &object, const ObjectID &object_id) {
+        RAY_LOG(DEBUG) << "Promoting object to plasma " << object_id;
+        bool object_exists;
+        RAY_RETURN_NOT_OK(plasma_store_provider_->Put(object, object_id, &object_exists));
```
Comment (ericl) [inline src/ray/core_worker/core_worker.cc:372]: Can we clean this up?

**Comment by ericl** [comment]
I believe the core C++ unit tests are failing.

File: python/ray/worker.py [hunk 2]
```diff
@@ -1280,14 +1280,6 @@ def connect(node,
         worker.core_worker.set_object_store_client_options(
             "ray_driver_{}".format(os.getpid()), driver_object_store_memory)
 
-    # Put something in the plasma store so that subsequent plasma store
-    # accesses will be faster. Currently the first access is always slow, and
-    # we don't want the user to experience this.
-    if mode != LOCAL_MODE:
-        temporary_object_id = ray.ObjectID.from_random()
-        worker.put_object(1, object_id=temporary_object_id)
-        ray.internal.free([temporary_object_id])
```
Comment (zhuohan123) [inline python/ray/worker.py:1289]: add this optimization back

**Comment by ericl** [comment]
Hey one more thing, can we have a feature flag for this (on by default)? Might also be an easy way to fix the java tests -- flag it off for those.

**Comment by ericl** [comment]
++++/home/travis/build/ray-project/tmp.P9tmJ5PmCc/ci/travis/check-git-clang-format-output.sh
Running clang-format against branch 8fcfcc41005c1bdf046ae6866101990db159d35a, with hash 8fcfcc41005c1bdf046ae6866101990db159d35a
clang-format failed:
diff --git a/src/ray/core_worker/cor

**Comment by ericl** [comment]
Seems like test_global_gc_when_full and test_object_transfer_retry might be failing, you can try running the locally to verify. Not sure about streaming tests.

**Comment by zhuohan123** [comment]
> Seems like test_global_gc_when_full and test_object_transfer_retry might be failing, you can try running the locally to verify. Not sure about streaming tests.

I believe this is caused by the master branch: https://travis-ci.com/github/ray-project/ray/jobs/357748807

**Comment by zhuohan123** [comment]
@ericl can you restart the mac tests? all tests passed on my laptop

**Review by ericl — APPROVED** [review]

**Pull request merged**